frb_rust/src/ffi/io.rs

```diff
@@ -31,3 +38,89 @@ mod tests {
         assert_eq!(ns, 567_123_000);
     }
 }
+
+/// # Safety
+///
+/// This function should never be called manually.
+#[no_mangle]
+pub unsafe extern "C" fn new_dart_opaque(handle: Dart_Handle) -> usize {
+    Dart_NewPersistentHandle_DL_Trampolined(handle) as _
+}
+
+/// # Safety
+///
+/// This function should never be called manually.
+#[no_mangle]
+pub unsafe extern "C" fn get_dart_object(ptr: usize) -> Dart_Handle {
+    Dart_HandleFromPersistent_DL_Trampolined(ptr as _)
+}
+
+/// # Safety
+///
+/// This function should never be called manually.
+#[no_mangle]
+pub unsafe extern "C" fn drop_dart_object(ptr: usize) {
+    Dart_DeletePersistentHandle_DL_Trampolined(ptr as _);
+}
+
+#[derive(Debug)]
+/// Option for correct drop.
+pub struct DartHandleWrap(Option<Dart_PersistentHandle>);
+
+impl DartHandleWrap {
+    pub fn from_raw(ptr: Dart_PersistentHandle) -> Self {
+        Self(Some(ptr))
+    }
+
+    pub fn into_raw(mut self) -> Dart_PersistentHandle {
+        self.0.take().unwrap()
+    }
+}
+
+impl Drop for DartHandleWrap {
+    fn drop(&mut self) {
+        if let Some(inner) = self.0 {
+            unsafe { Dart_DeletePersistentHandle_DL_Trampolined(inner) }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct DartOpaqueBase {
+    inner: DartHandleWrap,
+    drop_port: Option<MessagePort>,
+}
+
+impl DartOpaqueBase {
+    pub fn new(handle: Dart_PersistentHandle, port: MessagePort) -> Self {
+        Self {
+            inner: DartHandleWrap(Some(handle)),
+            drop_port: Some(port),
+        }
+    }
+
+    /// Creates a new wrapper for send to dart.
+    ///
+    /// # Safety
+    ///
+    /// Must be used for Dart objects created on the current thread
+    /// only to be sent to the Dart side.
+    pub unsafe fn new_non_dropable(handle: Dart_Handle) -> Self {
```

Comment:
Btw is Dart_Handle same as Dart_PersistentHandle? If so maybe we should not mix them to make it clear

Reply:
Dart_Handle means the object can be cleaned up by the GC while Dart_Persistent_Handle is not.

Reply:
Btw where is new_non_dropable used? I have done a quick search but there is no result

![image](https://user-images.githubusercontent.com/5236035/205248250-e7e36d64-c1da-4ff1-b997-0b86caeeb26d.png)

Reply:
not used in tests. this function is for cases when we call dart from rust and want to return the received Dart_Handle to dart via frb (for direct ffi dart compatibility)

Reply:
Hmm, then maybe we should add a test, otherwise it can regress in the future

Reply:
Btw, how do we call Dart from Rust now? IMHO we can only do the reverse now

Reply:
It is possible to write registration of dart functions in rust.

Reply:
Sure, but how can we call it? IMHO we only have DartOpaque and RustOpaque, which is not callable

Reply:
"on rust side use function pointer" - looks like we have to execute some unsafe code?

It is quite interesting to see Rust can call Dart callbacks. We should test it and document it!

Reply:
new_non_dropable deleted

Reply:
Looking forward to the next PR!